File: sdk/nodejs/middleware.js
// Express middleware wrapper for Node.js SDK client

function expressRateLimit(client) {
  return async (req, res, next) => {
    try {
      const allowed = await client.verify(req.path);
      if (!allowed) {
        return res.status(429).json({
          error: 'Too Many Requests. Rate limit exceeded.'
        });
      }
      next();
    } catch (error) {
      // Fail-open strategy: allow requests on connection timeouts
      console.warn('[RateLimiter SDK] Gateway connection failed, failing open:', error.message);
      next();
    }
  };
}

module.exports = expressRateLimit;

File: sdk/go/middleware.go
package sdk

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// GinRateLimit returns a Gin middleware that rate limits endpoints using the SDK Client.
func GinRateLimit(client *Client) gin.HandlerFunc {
	return func(c *gin.Context) {
		// Use the request path as the verification route
		allowed, err := client.Verify(c.Request.Context(), c.Request.URL.Path)
		if err != nil {
			// Fail-open: If the rate limiting service is down, allow request but log warning
			c.Next()
			return
		}

		if !allowed {
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"error": "Too Many Requests. Rate limit exceeded.",
			})
			return
		}

		c.Next()
	}
}
